### DISCUSSION TITLE
failed to load external entity "Utilities/merge.xslt"

### DISCUSSION THREAD

**Discussion by MatthewYT117** [kinkofer/FightClub5eXML#174] [Q&A] [answered]
Hello, I am following the previous discussion #170 and am experiencing some issues

I copied the required dlls and exes as per the guide. Running xsltproc -version in cmd returns:
Using libxml 20708, libxslt 10126 and libexslt 815
xsltproc was compiled against libxml 20706, libxslt 10126 and libexslt 815
libxslt 10126 was compiled against libxml 20706
libexslt 815 was compiled against libxml 20706

I downloaded and unzipped the data to D:\FightClub5eXML_Master

This folder contains:
-Collections
-FightClub5eXML
-Utilities
-.editorconfig
-.gitignore
-.travis.yml
-LICENSE
-README.md

D:\FightClub5eXML_Master\FightClub5eXML contains:
-Sources
-CHANGELOG.md
-README.md

This is the folder where I created an xml called Collection.xml and it contains the same text as your example in the readme:

`<?xml version="1.0" encoding="UTF-8"?>
<collection>
    <doc href="../FightClub5eXML/Sources/PlayersHandbook.xml" />
    <doc href="../FightClub5eXML/Sources/DungeonMastersGuide.xml" />
    <doc href="../FightClub5eXML/Sources/MonsterManual.xml" />
</collection>`

(Sorry. I couldn't get this to preview correctly. But it is a copy and paste of the xml from the readme)

Within cmd, I navigated to D:\FightClub5eXML_Master\FightClub5eXML and ran this command:

xsltproc -o BuiltCollection.xml Utilities/merge.xslt Collection.xml

and returned this message:
warning: failed to load external entity "Utilities/merge.xslt"
cannot parse Utilities/merge.xslt

I also copied the Complete.xml from your provided Collections folder to try that and changed the command line input to 

xsltproc -o BuiltCollection.xml Utilities/merge.xslt Complete.xml

and it returned the same message as above

Very much a newbie similar to the user in the previous discussion. Is there any assistance you can provide to get it working?

Many thanks

**Comment by MatthewYT117** [comment, marked as answer]
I changed my input command to contain the full path:

xsltproc -o D:/FightClub5eXML_Master/FightClub5eXML/BuiltCollection.xml D:/FightClub5eXML_Master/Utilities/merge.xslt D:/FightClub5eXML_Master/FightClub5eXML/Collection.xml

I then recieved this error message:

warning: failed to load external entity "/FightClub5eXML_Master/FightClub5eXML/Sources/PlayersHandbook.xml"
warning: failed to load external entity "/FightClub5eXML_Master/FightClub5eXML/Sources/DungeonMastersGuide.xml"
warning: failed to load external entity "/FightClub5eXML_Master/FightClub5eXML/Sources/MonsterManual.xml"
warning: failed to load external entity "/FightClub5eXML_Master/FightClub5eXML/Sources/PlayersHandbook.xml"
warning: failed to load external entity "/FightClub5eXML_Master/FightClub5eXML/Sources/DungeonMastersGuide.xml"
warning: failed to load external entity "/FightClub5eXML_Master/FightClub5eXML/Sources/MonsterManual.xml"
warning: failed to load external entity "/FightClub5eXML_Master/FightClub5eXML/Sources/PlayersHandbook.xml"
warning: failed to load external entity "/FightClub5eXML_Master/FightClub5eXML/Sources/DungeonMastersGuide.xml"
warning: failed to load external entity "/FightClub5eXML_Master/FightClub5eXML/Sources/MonsterManual.xml"

I took a look in that folder and those .xml files don't exist.

I replaced the Collection.xml with the Complete.xml you provide and it seems to have worked correctly. I haven't imported it to the app just yet though.

Have I misunderstood the readme? Am I missing something from the example and that's why the initial command did not work?

Thanks

**Reply by zcdziura** [reply]
Ah, I see what your issue is.

Your original, custom Collection file was listed as this:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<collection>
    <doc href="../FightClub5eXML/Sources/PlayersHandbook.xml" />
    <doc href="../FightClub5eXML/Sources/DungeonMastersGuide.xml" /> 
    <doc href="../FightClub5eXML/Sources/MonsterManual.xml" />
</collection>
```
However, if you take a look through the `FightClub5eXML/Sources` folder, you won't see a list of XML files, you'll actually see a list of folders, each with its own series of XML files.

You'll have to amend your custom collection file to reference each of those smaller XML files.

The README at the root of the project should be updated to reflect that.

**Reply by MatthewYT117** [reply]
Thanks for that. I reran the tool with the CoreRulebooksAndSupplements.xml you provide and imported the collection this morning. 

Much appreciated that you have taken the time to put together this repository!

**Reply by cjengl** [reply]
I have been getting similar errors. I either get one error that says:
"Warning: failed to load external entity "Utilties/merge.xslt"" 
or a very long list with the same error but for each listing in the collection.xml for example "warning: failed to load external entity "../FightClub5eXML/Sources/PlayersHandbook/backgrounds-phb.xml". 

What folder should my Collection.xml file be in? What location should I have for my command prompt? Once I have them in place what's the script I should run in the command prompt to execute the merge?

**Reply by zcdziura** [reply]
I have my collection files stored in `C:\Users\MY_USER_NAME\Documents\FightClub5eXML`, though the collections can be in any folder anywhere on your system.

It is important you open your command prompt and navigate to that top-level folder before you execute any commands. So in my example, if I open up PowerShell, first I have to execute `cd "C:\Users\MY_USER_NAME\Documents\FightClub5eXML`.

The collection files use relative paths in order to point to all of the smaller data files; that's what the ".." implies in the file name. By convention, ".." is a pointer to the parent folder. So if you're in `C:\Users\MY_USER_NAME\Documents\FightClub5eXML`, ".." points to `C:\Users\MY_USER_NAME\Documents`.

That could be why `xsltproc` is having trouble executing the `merge.xslt` script.

**Reply by cjengl** [reply]
Thanks! My zipped file is FightClub5e-master so when I unzip it I should navigate to that folder? So it should say C:/users/me/documents/FightClub5e-master>  Then I execute the script? What's the script once I'm in that location and can I just use a notepad to copy paste the xml in the collections on the repository and save it as Collection.xml? Sorry never used xml before

**Reply by zcdziura** [reply]
I'll use one of the existing collections within this repository to detail what to do. Let's say you want to build a compendium from the "CoreOnly.xml" collection, found in the "Collections" folder. Do the following:

1. Download a ZIP archive of this repository (FightClub5e-master, as you have done) and extract it to your "My Documents" folder; `C:\Users\MY_USER_NAME\Documents\FightClub5e-master`. This location will be referred to as the "repository root folder".
2. Open up PowerShell and navigate to that folder; `cd C:\Users\MY_USER_NAME\Documents\FightClub5e-master`.
3. Within PowerShell, execute the following command: `xsltproc -o Compendium.xml Utilities/merge.xslt Collections/CoreOnly.xml`.

The above command will execute the `xsltproc` program, which will itself run the "merge.xslt" script found within the Utilities folder. That "merge.xslt" script takes in a file as an input, in this case the "CoreOnly.xml" file found within the Collections folder. The "merge.xslt" script will create a complete Compendium file, itself filled with all of the data listed in the "CoreOnly.xml" collection file, and will save that file as "Compendium.xml", as denoted by the `-o` flag passed to `xsltproc`.

The newly created "Compendium.xml" file will be found in the repository root folder, and from there you can copy it to your phone and import it into FightClub 5e.

I hope that clarifies things!

**Reply by cjengl** [reply]
Thank you it looks like it worked!
